src/group_theory/submonoid/inverses.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2021 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import group_theory.submonoid.operations
+
+/-!
+
+# Submonoid of inverses
+
+Given a submonoid `N` of a monoid `M`, we define the submonoid `N.left_inv` as the submonoid of
+left inverses of `N`. When `M` is commutative, we may define `from_comm_left_inv : N.left_inv →* N`
+since the inverses are unique. When `N ≤ is_unit.submonoid M`, this is precisely
+the pointwise inverse of `N`, and we may define `left_inv_equiv : S.left_inv ≃* S`.
+
+For the pointwise inverse of submonoids of groups, please refer to
+`group_theory.submonoid.pointwise`.
+
+## TODO
+
+Define the submonoid of right inverses and two-sided inverses.
```

Comment:
A neat way to generalize to right-inverses is:
```
/-- `S.smul_inv` is the submonoid containing all the smul-inverses of `S`. -/
def smul_inv (α : Type*) [monoid α] [mul_action M α] [is_scalar_tower M α α] : submonoid α :=
{ carrier := { x : α | ∃ y : S, y • x = 1 },
  one_mem' := ⟨1, one_smul _ _⟩,
  mul_mem' := λ a b ⟨a', ha⟩ ⟨b', hb⟩,
    ⟨b' * a', by rw [mul_smul, ←smul_mul_assoc, ha, one_mul, hb]⟩ }

/-- `S.left_inv` is the submonoid containing all the left inverses of `S`. -/
def left_inv {α : Type*} : submonoid M :=
S.smul_inv M

def to_opposite : submonoid Mᵐᵒᵖ :=
{ carrier := mul_opposite.unop ⁻¹' (S : set M),
  one_mem' := S.one_mem,
  mul_mem' := λ x y hx hy, S.mul_mem hy hx}

/-- `S.right_inv` is the submonoid containing all the right inverses of `S`. -/
def right_inv {α : Type*} : submonoid M :=
S.to_opposite.smul_inv M
```

I don't think we want to do that in the PR, but adding this comment here in case we want to come back to it later.

Reply:
I think we should add this code snippet in the TODO comment. Otherwise it will be lost forever.